### PR TITLE
Circuit: Add constraints

### PR DESCRIPTION
**Pull request by ConstanceBeguier**
Add the constraint: (split_flag=1) => (is_native_asset=0)

Replace the constraint: (v_old=0) or (root=anchor)
by the constraint: (v_old=0 and split_flag=0) or (root=anchor)

Limit the version of half (< 2.3) because recent half versions required at least rust version 1.70.

### CODE REVIEW

File: src/circuit.rs [hunk 1]
```diff
@@ -276,9 +277,14 @@ impl plonk::Circuit<pallas::Base> for Circuit {
                             - v_new.clone()
                             - magnitude * sign,
                     ),
+                    // We already checked that
+                    // * split_flag is boolean (just above), and
+                    // * v_old is a 64 bit integer (in note commitment evaluation).
+                    // So, split_flag + v_old = 0 only when both split_flag and v_old are equal to
+                    // zero (no overflow can occur).
```
Comment (PaulLaux) [inline src/circuit.rs:284]: So, split_flag + v_old = 0 only when (split_flag = 0 and v_old = 0), no overflow can occur.

File: src/circuit.rs [hunk 2]
```diff
@@ -276,9 +277,14 @@ impl plonk::Circuit<pallas::Base> for Circuit {
                             - v_new.clone()
                             - magnitude * sign,
                     ),
+                    // We already checked that
+                    // * split_flag is boolean (just above), and
+                    // * v_old is a 64 bit integer (in note commitment evaluation).
```
Comment (PaulLaux) [inline src/circuit.rs:282]: (in the note commitment evaluation)

File: src/circuit.rs [hunk 3]
```diff
@@ -1550,7 +1560,9 @@ mod tests {
                 let (circuit, instance) =
                     generate_circuit_instance(is_native_asset, split_flag, &mut rng);
 
-                check_proof_of_orchard_circuit(&circuit, &instance, true);
+                let should_pass = !matches!((is_native_asset, split_flag), (true, true));
```
Comment (PaulLaux) [inline src/circuit.rs:1563]: very cool, let's emphasize the not:
`!(matches!((is_native_asset, split_flag), (true, true)));`

File: src/circuit.rs [hunk 4]
```diff
@@ -1619,19 +1631,47 @@ mod tests {
                 };
                 check_proof_of_orchard_circuit(&circuit, &instance_wrong_cmx_pub, false);
 
-                // If split_flag=0, set nf_old_pub to random Nullifier
+                // Set nf_old_pub to random Nullifier
```
Comment (PaulLaux) [inline src/circuit.rs:1634]: to be a random Nullifier

**Review by PaulLaux — APPROVED** [review]
Added some minor comments

File: src/circuit.rs (hunk 1 above)
Comment (ConstanceBeguier) [reply src/circuit.rs:284]: Done

File: src/circuit.rs (hunk 2 above)
Comment (ConstanceBeguier) [reply src/circuit.rs:282]: Done

File: src/circuit.rs (hunk 4 above)
Comment (ConstanceBeguier) [reply src/circuit.rs:1634]: Done

Comment (ConstanceBeguier) [reply src/circuit.rs]: Done

File: src/circuit.rs (hunk 3 above)
Comment (ConstanceBeguier) [reply src/circuit.rs]: Done

**Pull request merged**